File: 04. Support Vector Regressor (SVR).R
## Support Vector Regression ##

# Básicamente tiene la misma estructura que SVM, pero que empleado para predecir
# en lugar de clasificar

# libraries ---------------------------------------------------------------

library(tidyverse)
library(e1071)


# read data ---------------------------------------------------------------

position_salaries <- readr::read_csv("data/Position_Salaries.csv") %>% 
  janitor::clean_names()

position_salaries <- position_salaries %>% 
  dplyr::select(-position)

# ajuste del modelo -------------------------------------------------------

regresion <- position_salaries %>% 
  e1071::svm(salary ~ ., 
             type = "eps-regression", # especificar que es una regresión
             kernel = "radial", # el problema no es lineal
             data = .)

# Visualización -----------------------------------------------------------

## Los valores atípicos suelen ser penalizados por las SVM, por eso el
## nivel 10 no se predice "tan bien" como deberia

position_salaries %>% 
  dplyr::mutate(ajustados = regresion$fitted) %>% 
  ggplot2::ggplot(aes(level, y = salary)) + 
  geom_point(color = "darkred") +
  geom_line(aes(y = ajustados), color = "darkblue") +
  labs(title = "SVR",
       x = "Nivel del empleado",
       y = "Salario ($)") +
  theme_classic()

# prediccion --------------------------------------------------------------

predict(regresion, data.frame(level = 6.5))

File: 04. Support Vector Regressor (SVR).py
# -*- coding: utf-8 -*-
"""
Created on Tue Aug  3 18:40:28 2021

@author: Admin
"""

#SVR

# =============================================================================
# Libraries
# =============================================================================

import pandas as pd
#import numpy as np
import matplotlib.pyplot as plt
import sklearn.preprocessing as sklp
import sklearn.svm as svm

# =============================================================================
# Read data
# =============================================================================

position_salaries = pd.read_csv("data\Position_Salaries.csv")

x = position_salaries.iloc[:, 1:2].values # es mejor que siempre sea matriz
y = position_salaries.iloc[:, 2].values

# =============================================================================
# Escalado de variables
# =============================================================================

## Escalar las variables es importante para este algoritmo

scl_x = sklp.StandardScaler()
scl_y = sklp.StandardScaler()

x = scl_x.fit_transform(x)
y = scl_y.fit_transform(y.reshape(-1, 1))

# =============================================================================
# Ajustar SVR
# =============================================================================

regresion = svm.SVR(kernel = "rbf") # Kernel gaussiano
regresion.fit(x, y)

# =============================================================================
# Visualizacion
# =============================================================================

## Nótese que se invierte la transformación!

plt.scatter(scl_x.inverse_transform(x), scl_y.inverse_transform(y), 
            color = "darkred")
plt.plot(scl_x.inverse_transform(x), 
         scl_y.inverse_transform(regresion.predict(x)), 
         color = "darkblue")
plt.title("Modelo SVR")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo en $")
plt.show()

# =============================================================================
# Prediccion
# =============================================================================

scl_y.inverse_transform(regresion.predict(scl_x.transform([[6.5]])))
